infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -172,10 +255,58 @@ void processOpaqueEntry(List<SMREntry> smrEntries, Long currentSeqNum, UUID dstU
         log.debug("Process the entries {} and set sequence number {}", smrEntries, currentSeqNum);
     }
 
+    /**
+     * Used by Full Snapshot Sync Phase II:
+     * Write a list of SMR entries to the specified real stream log.
+     * @param smrEntries
+     * @param currentSeqNum
+     * @param dstUUID
+     */
+    void processShadowOpaqueEntry(List<SMREntry> smrEntries, Long currentSeqNum, UUID dstUUID) {
+        CorfuStoreMetadata.Timestamp timestamp = logReplicationMetadataManager.getTimestamp();
+        long persistConfigID = logReplicationMetadataManager.query(timestamp, LogReplicationMetadataManager.LogReplicationMetadataType.TOPOLOGY_CONFIG_ID);
+        long persistSnapStart = logReplicationMetadataManager.query(timestamp, LogReplicationMetadataManager.LogReplicationMetadataType.LAST_SNAPSHOT_STARTED);
+        long persitSeqNum = logReplicationMetadataManager.query(timestamp, LogReplicationMetadataManager.LogReplicationMetadataType.LAST_SNAPSHOT_APPLIED_SEQ_NUM);
+
+        // Verify that the operation is in the correct order. If it is out of sync with the metadata, skip it.
+        if (siteConfigID != persistConfigID || srcGlobalSnapshot != persistSnapStart || currentSeqNum != (persitSeqNum + 1)) {
+            log.warn("Skip current processing as the persistent metadata {} has applied more recent message {} than current siteConfigID {} " +
+                    "srcGlobalSnapshot {} currentAppliedSeqNum {}", logReplicationMetadataManager.toString(),
+                    siteConfigID, srcGlobalSnapshot, currentSeqNum);
+            return;
+        }
+
+        TxBuilder txBuilder = logReplicationMetadataManager.getTxBuilder();
+
+        // Update the siteConfigID with the same value to fence off any metadata updates based on the same timestamp.
+        logReplicationMetadataManager.appendUpdate(txBuilder, LogReplicationMetadataManager.LogReplicationMetadataType.TOPOLOGY_CONFIG_ID, siteConfigID);
+
+        // Update the appliedSeqNumber.
+        logReplicationMetadataManager.appendUpdate(txBuilder, LogReplicationMetadataManager.LogReplicationMetadataType.LAST_SNAPSHOT_APPLIED_SEQ_NUM, currentSeqNum);
+
+        // Append log entries.
+        for (SMREntry smrEntry : smrEntries) {
+            txBuilder.logUpdate(dstUUID, smrEntry);
+        }
+
+        try {
+            txBuilder.commit(timestamp);
+        } catch (Exception e) {
+            log.warn("Caught an exception ", e);
```

Comment:
not all commits() need to rethrow an exception, do you want to throw the exception only if it is an UnrecoverableException